src/daft-core/src/utils/sketch.rs

```diff
@@ -0,0 +1,51 @@
+use common_error::DaftResult;
+use sketches_ddsketch::{Config, DDSketch};
+
+pub struct Sketch {
+    sketch: DDSketch,
+}
+
+impl Sketch {
+    pub fn new() -> Self {
+        Sketch {
+            sketch: DDSketch::new(Config::defaults()),
+        }
+    }
+
+    pub fn add(&mut self, value: f64) -> &mut Self {
+        self.sketch.add(value);
+        self
+    }
+
+    pub fn merge(&mut self, other: &Sketch) -> DaftResult<&mut Self> {
+        self.sketch.merge(&other.sketch)?;
+        Ok(self)
+    }
+
+    pub fn quantile(&self, q: f64) -> DaftResult<Option<f64>> {
+        Ok(self.sketch.quantile(q)?)
+    }
+
+    pub fn from_value(value: f64) -> Self {
+        let mut sketch = Sketch::new();
+        sketch.add(value);
+        sketch
+    }
+
+    pub fn from_binary(binary: &[u8]) -> DaftResult<Self> {
+        let sketch_str = std::str::from_utf8(binary)?;
+        let sketch: DDSketch = serde_json::from_str(sketch_str)?;
+        Ok(Sketch { sketch })
+    }
+
+    pub fn to_binary(&self) -> DaftResult<Vec<u8>> {
+        let sketch_str = serde_json::to_string(&self.sketch)?;
```

Comment:
`serde_bytes` might be a better choice here, since we don't need human-readability: https://docs.rs/serde_bytes/latest/serde_bytes/

Reply:
I think that is fine for now. I'm also experimenting with `serde_arrow` (https://docs.rs/serde_arrow/latest/serde_arrow/) which could actually potentially help us serialize `Vec<DDSketch>` directly into an Arrow structarray!

Reply:
Oh very fun! I was able to get DDSketch to serialize into Arrow arrays:

```rust
    use sketches_ddsketch::DDSketch;
    use serde_arrow::schema::{TracingOptions, SerdeArrowSchema, SchemaLike};

    let sketch1 = DDSketch::default();
    let mut sketch2 = DDSketch::default();
    for i in (1 .. 10) {
        sketch2.add(i as f64)
    }
    let records = vec![
        sketch1,
        sketch2,
    ];

    let fields = SerdeArrowSchema::from_type::<DDSketch>(TracingOptions::default()).unwrap().to_arrow2_fields().unwrap();
    let arrays = serde_arrow::to_arrow2(&fields, &records).unwrap();

    println!("Fields: {fields:?}\n");
    println!("Arrays: {arrays:?}\n");
```
This actually seems to work pretty well! When the code above is run, I get:

```
Fields: [Field { name: "config", data_type: Struct([Field { name: "max_num_bins", data_type: UInt32, is_nullable: false, metadata: {} }, Field { name: "gamma", data_type: Float64, is_nullable: false, metadata: {} }, Field { name: "gamma_ln", data_type: Float64, is_nullable: false, metadata: {} }, Field { name: "min_value", data_type: Float64, is_nullable: false, metadata: {} }, Field { name: "offset", data_type: Int32, is_nullable: false, metadata: {} }]), is_nullable: false, metadata: {} }, Field { name: "store", data_type: Struct([Field { name: "bins", data_type: LargeList(Field { name: "element", data_type: UInt64, is_nullable: false, metadata: {} }), is_nullable: false, metadata: {} }, Field { name: "count", data_type: UInt64, is_nullable: false, metadata: {} }, Field { name: "min_key", data_type: Int32, is_nullable: false, metadata: {} }, Field { name: "max_key", data_type: Int32, is_nullable: false, metadata: {} }, Field { name: "offset", data_type: Int32, is_nullable: false, metadata: {} }, Field { name: "bin_limit", data_type: UInt64, is_nullable: false, metadata: {} }, Field { name: "is_collapsed", data_type: Boolean, is_nullable: false, metadata: {} }]), is_nullable: false, metadata: {} }, Field { name: "negative_store", data_type: Struct([Field { name: "bins", data_type: LargeList(Field { name: "element", data_type: UInt64, is_nullable: false, metadata: {} }), is_nullable: false, metadata: {} }, Field { name: "count", data_type: UInt64, is_nullable: false, metadata: {} }, Field { name: "min_key", data_type: Int32, is_nullable: false, metadata: {} }, Field { name: "max_key", data_type: Int32, is_nullable: false, metadata: {} }, Field { name: "offset", data_type: Int32, is_nullable: false, metadata: {} }, Field { name: "bin_limit", data_type: UInt64, is_nullable: false, metadata: {} }, Field { name: "is_collapsed", data_type: Boolean, is_nullable: false, metadata: {} }]), is_nullable: false, metadata: {} }, Field { name: "min", data_type: Float64, is_nullable: false, metadata: {} }, Field { name: "max", data_type: Float64, is_nullable: false, metadata: {} }, Field { name: "sum", data_type: Float64, is_nullable: false, metadata: {} }, Field { name: "zero_count", data_type: UInt64, is_nullable: false, metadata: {} }]

Arrays: [StructArray[{max_num_bins: 2048, gamma: 1.02020202020202, gamma_ln: 0.020000666706669525, min_value: 0.000000001, offset: 1037}, {max_num_bins: 2048, gamma: 1.02020202020202, gamma_ln: 0.020000666706669525, min_value: 0.000000001, offset: 1037}], StructArray[{bins: [], count: 0, min_key: 2147483647, max_key: -2147483648, offset: 0, bin_limit: 2048, is_collapsed: false}, {bins: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0], count: 9, min_key: 0, max_key: 110, offset: -12, bin_limit: 2048, is_collapsed: false}], StructArray[{bins: [], count: 0, min_key: 2147483647, max_key: -2147483648, offset: 0, bin_limit: 2048, is_collapsed: false}, {bins: [], count: 0, min_key: 2147483647, max_key: -2147483648, offset: 0, bin_limit: 2048, is_collapsed: false}], Float64[inf, 1], Float64[-inf, 9], Float64[0, 45], UInt64[0, 0]]
```

Reply:
@maxime-petitjean we managed to make it work!

https://github.com/Eventual-Inc/Daft/pull/2090

If you rebase on `main`, you should now have access to a `daft_sketch::from_arrow`, `daft_sketch::into_arrow` and `daft_sketch::ARROW2_DDSKETCH_DTYPE`, which should help with serde of your `DDSketch` struct into arrow datastructures.

Reply:
I tried to use it without success because of some schema mismatch.
It looks like `daft_sketch::into_arrow2` generates data with `is_nullable` to false which is incompatible with `ARROW2_DDSKETCH_DTYPE`.

When executing this code (`sketch` is an `Option<DDSketch>`):
```rust
let arrow_array = daft_sketch::into_arrow2(vec![sketch]);
let datatype = DataType::from(&daft_sketch::ARROW2_DDSKETCH_DTYPE.clone());
DataArray::new(Field::new(&self.field.name, datatype).into(), arrow_array)
```
I get:
```
pyo3_runtime.PanicException: expected
  Struct([
    Field {
      name: "config",
      data_type: Struct([
        Field { name: "max_num_bins", data_type: UInt32, is_nullable: true, metadata: {} },
        Field { name: "gamma", data_type: Float64, is_nullable: true, metadata: {} },
        Field { name: "gamma_ln", data_type: Float64, is_nullable: true, metadata: {} },
        Field { name: "min_value", data_type: Float64, is_nullable: true, metadata: {} },
        Field { name: "offset", data_type: Int32, is_nullable: true, metadata: {} }
      ]),
      is_nullable: true,
      metadata: {}
    },
    Field {
      name: "store",
      data_type: Struct([
        Field { name: "bins", data_type: LargeList(Field { name: "item", data_type: UInt64, is_nullable: true, metadata: {} }), is_nullable: true, metadata: {} },
        Field { name: "count", data_type: UInt64, is_nullable: true, metadata: {} },
        Field { name: "min_key", data_type: Int32, is_nullable: true, metadata: {} },
        Field { name: "max_key", data_type: Int32, is_nullable: true, metadata: {} },
        Field { name: "offset", data_type: Int32, is_nullable: true, metadata: {} },
        Field { name: "bin_limit", data_type: UInt64, is_nullable: true, metadata: {} },
        Field { name: "is_collapsed", data_type: Boolean, is_nullable: true, metadata: {} }
      ]),
      is_nullable: true,
      metadata: {}
    },
    Field { name: "negative_store",
      data_type: Struct([
        Field { name: "bins", data_type: LargeList(Field { name: "item", data_type: UInt64, is_nullable: true, metadata: {} }), is_nullable: true, metadata: {} },
        Field { name: "count", data_type: UInt64, is_nullable: true, metadata: {} },
        Field { name: "min_key", data_type: Int32, is_nullable: true, metadata: {} },
        Field { name: "max_key", data_type: Int32, is_nullable: true, metadata: {} },
        Field { name: "offset", data_type: Int32, is_nullable: true, metadata: {} },
        Field { name: "bin_limit", data_type: UInt64, is_nullable: true, metadata: {} },
        Field { name: "is_collapsed", data_type: Boolean, is_nullable: true, metadata: {} }
      ]),
      is_nullable: true,
      metadata: {}
    },
    Field { name: "min", data_type: Float64, is_nullable: true, metadata: {} },
    Field { name: "max", data_type: Float64, is_nullable: true, metadata: {} },
    Field { name: "sum", data_type: Float64, is_nullable: true, metadata: {} },
    Field { name: "zero_count", data_type: UInt64, is_nullable: true, metadata: {} }
  ])
  
, got

  Struct([
    Field {
      name: "config",
      data_type: Struct([
        Field { name: "max_num_bins", data_type: UInt32, is_nullable: false, metadata: {} },
        Field { name: "gamma", data_type: Float64, is_nullable: false, metadata: {} },
        Field { name: "gamma_ln", data_type: Float64, is_nullable: false, metadata: {} },
        Field { name: "min_value", data_type: Float64, is_nullable: false, metadata: {} },
        Field { name: "offset", data_type: Int32, is_nullable: false, metadata: {} }
      ]),
      is_nullable: false,
      metadata: {}
    },
    Field {
      name: "store",
      data_type: Struct([
        Field { name: "bins", data_type: LargeList(Field { name: "element", data_type: UInt64, is_nullable: false, metadata: {} }), is_nullable: false, metadata: {} },
        Field { name: "count", data_type: UInt64, is_nullable: false, metadata: {} },
        Field { name: "min_key", data_type: Int32, is_nullable: false, metadata: {} },
        Field { name: "max_key", data_type: Int32, is_nullable: false, metadata: {} },
        Field { name: "offset", data_type: Int32, is_nullable: false, metadata: {} },
        Field { name: "bin_limit", data_type: UInt64, is_nullable: false, metadata: {} },
        Field { name: "is_collapsed", data_type: Boolean, is_nullable: false, metadata: {} }
      ]),
      is_nullable: false,
      metadata: {}
    },
    Field {
      name: "negative_store",
      data_type: Struct([
        Field { name: "bins", data_type: LargeList(Field { name: "element", data_type: UInt64, is_nullable: false, metadata: {} }), is_nullable: false, metadata: {} },
        Field { name: "count", data_type: UInt64, is_nullable: false, metadata: {} },
        Field { name: "min_key", data_type: Int32, is_nullable: false, metadata: {} },
        Field { name: "max_key", data_type: Int32, is_nullable: false, metadata: {} },
        Field { name: "offset", data_type: Int32, is_nullable: false, metadata: {} },
        Field { name: "bin_limit", data_type: UInt64, is_nullable: false, metadata: {} },
        Field { name: "is_collapsed", data_type: Boolean, is_nullable: false, metadata: {} }
      ]),
      is_nullable: false,
      metadata: {}
    },
    Field { name: "min", data_type: Float64, is_nullable: false, metadata: {} },
    Field { name: "max", data_type: Float64, is_nullable: false, metadata: {} },
    Field { name: "sum", data_type: Float64, is_nullable: false, metadata: {} },
    Field { name: "zero_count", data_type: UInt64, is_nullable: false, metadata: {} }
  ])
  
when creating a new DataArray
```

Reply:
Ah - yes, our `DataArray` is only meant to hold primitive types (i64, strings etc). We don't currently use it to house Struct types!

`Series` with struct types are instead backed by our own `StructArray`. Here's how you can create a `StructArray` from arrow2 arrays: https://github.com/Eventual-Inc/Daft/blob/main/src/daft-core/src/array/ops/from_arrow.rs#L125-L149

```rust
    let sketches = vec![Some(DDSketch::new(Config::default())), None];

    // This also works (going directly to a Series)
    // let series = Series::from_arrow(Arc::new(Field::new("foo", DataType::from(&*daft_sketch::ARROW2_DDSKETCH_DTYPE))), daft_sketch::into_arrow2(sketches)).unwrap();
    // println!("Series: {}", series);

    let struct_array: StructArray = StructArray::from_arrow(Arc::new(Field::new("foo", DataType::from(&*daft_sketch::ARROW2_DDSKETCH_DTYPE))), daft_sketch::into_arrow2(sketches)).unwrap();
    println!("DataArray: {:?}", struct_array);
    let series = struct_array.into_series();

    let roundtripped = daft_sketch::from_arrow2(series.to_arrow()).unwrap();

    println!("Roundtripped vec: {:?}", roundtripped.len());
```

That is to say, we will probably have to change your code to implement kernels over `StructArray` instead of `BinaryArray`!

Reply:
It works with `StructArray`, thx!

I just pushed . There is still work to do (array of percentiles for `approx_percentile` parameter, clean up, ...) but it shows that your work is very useful!